lib/executor/variable_looping_vus.go

```diff
@@ -0,0 +1,620 @@
+// their last iterations is pretty simple. It just traverses the raw execution
+// steps and whenever there's a scaling down of VUs, it prevents the number of
+// VUs from decreasing for the configured gracefulRampDown period.
+//
+// Finishing up the test, i.e. making sure we have a step with 0 VUs at time
+// executorEndOffset, is not handled here. Instead GetExecutionRequirements()
+// takes care of that. But to make its job easier, this method won't add any
+// steps with an offset that's greater or equal to executorEndOffset.
+func (vlvc VariableLoopingVUsConfig) reserveVUsForGracefulRampDowns( //nolint:funlen
+	rawSteps []lib.ExecutionStep, executorEndOffset time.Duration,
+) []lib.ExecutionStep {
+	rawStepsLen := len(rawSteps)
+	gracefulRampDownPeriod := vlvc.GetGracefulRampDown()
+	newSteps := []lib.ExecutionStep{}
+
+	lastPlannedVUs := uint64(0)
+	for rawStepNum := 0; rawStepNum < rawStepsLen; rawStepNum++ {
+		rawStep := rawSteps[rawStepNum]
+		// Add the first step or any step where the number of planned VUs is
+		// greater than the ones in the previous step. We don't need to worry
+		// about reserving time for ramping-down VUs when the number of planned
+		// VUs is growing. That's because the gracefulRampDown period is a fixed
+		// value and any timeouts from early steps with fewer VUs will get
+		// overshadowed by timeouts from latter steps with more VUs.
+		if rawStepNum == 0 || rawStep.PlannedVUs > lastPlannedVUs {
+			newSteps = append(newSteps, rawStep)
+			lastPlannedVUs = rawStep.PlannedVUs
+			continue
+		}
+
+		// We simply skip steps with the same number of planned VUs
+		if rawStep.PlannedVUs == lastPlannedVUs {
+			continue
+		}
+
+		// If we're here, we have a downward "slope" - the lastPlannedVUs are
+		// more than the current rawStep's planned VUs. We're going to look
+		// forward in time (up to gracefulRampDown) and inspect the rawSteps.
+		// There are a 3 possibilities:
+		//  - We find a new step within the gracefulRampDown period which has
+		//    the same number of VUs or greater than lastPlannedVUs. Which
+		//    means that we can just advance rawStepNum to that number and we
+		//    don't need to worry about any of the raw steps in the middle!
+		//    Both their planned VUs and their gracefulRampDown periods will
+		//    be lower than what we're going to set from that new rawStep -
+		//    we've basically found a new upward slope or equal value again.
+		//  - We reach executorEndOffset, in which case we are done - we can't
+		//    add any new steps, since those will be after the executor end
+		//    offset.
+		//  - We reach the end of the rawSteps, or we don't find any higher or
+		//    equal steps to prevStep in the next gracefulRampDown period. So
+		//    we'll simply try to add an entry into newSteps with the values
+		//    {prevStep.TimeOffset + gracefulRampDown, rawStep.PlannedVUs} and
+		//    we'll continue with traversing the following rawSteps.
```

Comment:
Hey @na--, I'm looking into #1296, and was confused by this comment/functionality, as it seems to be the cause for at least the first variable-looping-vus bug.

So in the case of the example in that issue, with `gracefulRampDown: "5s"`, and each stage configured with a duration of 20s (for ease of calculating, since each step is exactly 1s apart), we have the following execution plan before it's modified by `VariableLoopingVUsConfig.reserveVUsForGracefulRampDowns()` (I shortened the output, but you get the idea; `1000000000` is 1s, etc.):

```golang
[]github.com/loadimpact/k6/lib.ExecutionStep len: 41, cap: 64, [
        {TimeOffset: 0, PlannedVUs: 0, MaxUnplannedVUs: 0},
        {TimeOffset: 1000000000, PlannedVUs: 1, MaxUnplannedVUs: 0},
        {TimeOffset: 2000000000, PlannedVUs: 2, MaxUnplannedVUs: 0},
        {TimeOffset: 3000000000, PlannedVUs: 3, MaxUnplannedVUs: 0},
        ...
        {TimeOffset: 18000000000, PlannedVUs: 18, MaxUnplannedVUs: 0},
        {TimeOffset: 19000000000, PlannedVUs: 19, MaxUnplannedVUs: 0},
        {TimeOffset: 20000000000, PlannedVUs: 20, MaxUnplannedVUs: 0},
        {TimeOffset: 21000000000, PlannedVUs: 19, MaxUnplannedVUs: 0},
        {TimeOffset: 22000000000, PlannedVUs: 18, MaxUnplannedVUs: 0},
        {TimeOffset: 23000000000, PlannedVUs: 17, MaxUnplannedVUs: 0},
        ...
        {TimeOffset: 38000000000, PlannedVUs: 2, MaxUnplannedVUs: 0},
        {TimeOffset: 39000000000, PlannedVUs: 1, MaxUnplannedVUs: 0},
        {TimeOffset: 40000000000, PlannedVUs: 0, MaxUnplannedVUs: 0},
]
```

And here's what is returned by `VariableLoopingVUsConfig.reserveVUsForGracefulRampDowns()`:

```golang
[]github.com/loadimpact/k6/lib.ExecutionStep len: 35, cap: 64, [
        {TimeOffset: 0, PlannedVUs: 0, MaxUnplannedVUs: 0},
        {TimeOffset: 1000000000, PlannedVUs: 1, MaxUnplannedVUs: 0},
        {TimeOffset: 2000000000, PlannedVUs: 2, MaxUnplannedVUs: 0},
        {TimeOffset: 3000000000, PlannedVUs: 3, MaxUnplannedVUs: 0},
        ...
        {TimeOffset: 18000000000, PlannedVUs: 18, MaxUnplannedVUs: 0},
        {TimeOffset: 19000000000, PlannedVUs: 19, MaxUnplannedVUs: 0},
        {TimeOffset: 20000000000, PlannedVUs: 20, MaxUnplannedVUs: 0},
        {TimeOffset: 26000000000, PlannedVUs: 19, MaxUnplannedVUs: 0},
        {TimeOffset: 27000000000, PlannedVUs: 18, MaxUnplannedVUs: 0},
        {TimeOffset: 28000000000, PlannedVUs: 17, MaxUnplannedVUs: 0},
        {TimeOffset: 29000000000, PlannedVUs: 16, MaxUnplannedVUs: 0},
        ...
        {TimeOffset: 37000000000, PlannedVUs: 8, MaxUnplannedVUs: 0},
        {TimeOffset: 38000000000, PlannedVUs: 7, MaxUnplannedVUs: 0},
        {TimeOffset: 39000000000, PlannedVUs: 6, MaxUnplannedVUs: 0},
]
```

Notice the jump from 20s/20VUs to 26s/19VUs and decreasing VUs and increasing TimeOffset, but always with a 6s delta after the first stage (because of the 5s `gracefulRampDown`).

So the reason for the jumpiness seen in the first bug at t=16s (in script exec. time) is that the `select` in `handleAllRawSteps` will apply steps from both `rawStepEvents` and `gracefulLimitEvents` at close to the same `TimeOffset` yet with conflicting `PlannedVUs` values.

I guess my question is what should the correct behavior be in this case? I added a second test case with this specific scenario in mind at https://github.com/imiric/k6/commit/9e39de6a88216c , but I'm still confused by the results.

Calculating the optimal number of VUs at any point in time doesn't seem reliable with this "check next steps if on a downward slope and add a fixed gracefulRampdown amount of offset" approach, but I haven't thought much about an alternative. What do you think?

Reply:
> Calculating the optimal number of VUs at any point in time doesn't seem reliable with this "check next steps if on a downward slope and add a fixed gracefulRampdown amount of offset" approach, but I haven't thought much about an alternative. What do you think?

:confused: both of the execution plans in your example seem correct to me. Am I missing something? 

Remember, these execution plans are meant to represent how many VUs the executor will require at any given point in time, with `gracefulRampDown: 0` and with whatever value the user specified, respectively. Then the `Run()` method tries to cleverly use these 2 potentially different plans to schedule VUs and to gracefully or completely shut them down, exactly when it has to. So, in this case, it seems to me like the bug is in how we interpret these execution requirements/plans in the `Run()` method... 

This check specifically seems wrong to me: `if step.PlannedVUs > currentScheduledVUs {`. I think it should be `if step.PlannedVUs > currentMaxAllowedVUs {`, and maybe moved in the `handleNewMaxAllowedVUs()` lambda

Reply:
Hhmm you're right, https://github.com/imiric/k6/commit/e3dc53c5c306fa09 appears to have fixed the incorrect starts at t=16s.

Still, I'm confused by the second execution plan streamed by `gracefulLimitEvents`. Let's discuss it at today's meeting.